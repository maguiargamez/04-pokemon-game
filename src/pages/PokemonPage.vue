<template>
  <div>

    <h1 v-if="!pokemon">Espere por favor...</h1>
    <div v-else>
      <h1>¿Quién es este pokémon?</h1>
      <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/>
      
      <PokemonOptions 
        :pokemons= "pokemonArr"
        @selectionPokemon="checkAnswer"
      />   

      <template v-if="showAnswer">
        <h2 class="fade-in">{{ message }}</h2> 
        <button @click="newGame">
          Nuevo juego
        </button>       
      </template>


    </div>



  </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture"
import PokemonOptions from "@/components/PokemonOptions"
import getPokemonOptions from '@/helpers/getPokemonOptions'

// getPokemonOptions()

export default {
  name: "PokemonPage",
  components: {
    PokemonPicture,
    PokemonOptions,
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions()
      const rndInt = Math.floor( Math.random() * 4 )
      this.pokemon = this.pokemonArr[ rndInt ]
    },
    checkAnswer( selectedId ){
      this.showPokemon= true
      this.showAnswer= true
      if( selectedId === this.pokemon.id ){
        this.message= `Correcto es ${ this.pokemon.name }!!!!`
      }
      else{
        this.message= `Ooops, era ${ this.pokemon.name } =(`
      }
    },
    newGame(){
      this.pokemonArr= []
      this.pokemon= null,
      this.showPokemon= false      
      this.showAnswer= false
      this.message= ''
      this.mixPokemonArray()
    }
  },
  mounted(){
    this.mixPokemonArray()
  }
};
</script>
