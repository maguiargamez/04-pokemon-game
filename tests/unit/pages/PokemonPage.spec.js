import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage component', ()=>{ 
    let wrapper

    beforeEach(()=>{
        wrapper = shallowMount( PokemonPage )
    })

    test('Debe de hacer match con el snapshot', ()=>{
        //console.log(wrapper.html())
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe de llamar el mixPokemonArray al montar', ()=>{
        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )
        const wrapper = shallowMount( PokemonPage )

        expect( mixPokemonArraySpy ).toHaveBeenCalled()
    })

    test('debe de hacer match con el snapshot cuando cargan los pokemon', ()=>{
        const wrapper = mount( PokemonPage, {
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        } )

        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', ()=>{

        const wrapper = shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        } )

        //console.log(wrapper.html())

        
        const pokemonPictureComponent = wrapper.find('pokemon-picture-stub')
        const pokemonOptionsComponent = wrapper.find('pokemon-options-stub')

        //PokemonPicture debe de existir
        expect( pokemonPictureComponent.exists() ).toBeTruthy()        

        //PokemonOptions debe de existir        
        expect( pokemonOptionsComponent.exists() ).toBeTruthy() 

        //PokemonPicture attribute pokemonId === 1
        expect( pokemonPictureComponent.attributes('pokemonid') ).toBe('1')

        //PokemonOptions attribute pokemons toBe true
        expect( pokemonOptionsComponent.attributes('pokemons') ).toBeTruthy()  

    })

    test('prubeas con checkAnswer',async  ()=> {
        const wrapper = shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        } )

        await wrapper.vm.checkAnswer(1)

        expect( wrapper.find('h2').exists() ).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBeTruthy()
        expect( wrapper.find('h2').text() ).toBe(`Correcto es ${ pokemons[0].name }!!!!`)

        await wrapper.vm.checkAnswer(3)
        expect( wrapper.vm.message ).toBe(`Ooops, era ${ pokemons[0].name } =(`)
    })
})