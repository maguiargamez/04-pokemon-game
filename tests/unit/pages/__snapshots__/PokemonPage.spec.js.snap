// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component Debe de hacer match con el snapshot 1`] = `
<div>
  <h1>Espere por favor...</h1>
</div>
`;

exports[`PokemonPage component debe de hacer match con el snapshot cuando cargan los pokemon 1`] = `
<div>
  <div>
    <h1>¿Quién es este pokémon?</h1>
    <div>
      <div class="pokemon-container"><img src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg" class="hidden-pokemon" alt="pokemon"></div>
    </div>
    <div>
      <div class="options-container">
        <ul>
          <li>bulbasaur</li>
          <li>ivysaur</li>
          <li>venusaur</li>
          <li>charmander</li>
        </ul>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>
`;
